library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentsImpl.java

```diff
@@ -0,0 +1,110 @@
+import org.quiltmc.qsl.component.impl.util.ErrorUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@ApiStatus.Internal
+public class ComponentsImpl {
+	private static final Logger LOGGER = LoggerFactory.getLogger("QSL/Components");
+	public static final String MODID = "quilt_component";
+    private static final Map<ComponentInjectionPredicate, Set<Identifier>> INJECTION_REGISTRY = new HashMap<>();
+	private static final RegistryKey<Registry<ComponentType<?>>> REGISTRY_KEY =
+			RegistryKey.ofRegistry(new Identifier("quilt", "components"));
+	public static final Registry<ComponentType<?>> REGISTRY =
+			new SimpleRegistry<>(REGISTRY_KEY, Lifecycle.experimental(), null);
+
+	public static <C extends Component> void inject(ComponentInjectionPredicate predicate, ComponentType<C> id) {
+		if (REGISTRY.get(id.id()) == null) {
+			throw new IllegalArgumentException(
+					"The target id %s does not match any registered component!".formatted(id.toString())
+			);
+		}
+
+		if (INJECTION_REGISTRY.containsKey(predicate)) {
+			if (!INJECTION_REGISTRY.get(predicate).add(id.id())) {
+				throw new IllegalStateException(
+						"Cannot inject the predicate %s with %s more than once! Consider creating a new component!"
+								.formatted(predicate.toString(), id.id().toString())
+				);
+			}
+		} else {
+			INJECTION_REGISTRY.put(predicate, Util.make(new HashSet<>(), set -> set.add(id.id())));
+		}
+
+		ComponentInjectionCache.getInstance().clear(); // Always clear the cache after an injection is registered.
+	}
+
+	public static <T extends Component> ComponentType<T> register(Identifier id, Component.Factory<T> factory) {
+		var componentType = new ComponentType<>(id, factory, false);
+		return Registry.register(REGISTRY, id, componentType);
+	}
+
+	public static <C extends Component> ComponentType<C> registerStatic(Identifier id, Component.Factory<C> factory) {
+		var componentType = new ComponentType<>(id, factory, true);
+		return Registry.register(REGISTRY, id, componentType);
+	}
+
+	public static List<ComponentType<?>> getInjections(ComponentProvider provider) {
+		return ComponentInjectionCache.getInstance().getCache(provider.getClass()).orElseGet(() -> {
+			List<ComponentType<?>> injectedTypes = INJECTION_REGISTRY.entrySet().stream()
+					.filter(it -> it.getKey().canInject(provider))
+					.map(Map.Entry::getValue)
+					.flatMap(Set::stream)
+					.map(ComponentsImpl::getEntry)
+					.collect(Collectors.toList());
+
+			ComponentInjectionCache.getInstance().record(provider.getClass(), injectedTypes);
+
+			return injectedTypes;
+		});
+	}
+
+	public static ComponentType<?> getEntry(Identifier id) {
+		return REGISTRY.getOrEmpty(id).orElseThrow(
+				ErrorUtil.illegalArgument("Cannot access element with id %s in the component registry!".formatted(id))
+		);
+	}
+
+	@NotNull
+	public static Logger getLogger() {
```

Comment:
...why not just make the logger `public`?

Reply:
Just a thing I have been doing for all my mods. It looks better in code to me. If you insist thought, I have no issue just making the LOGGER field public.

Reply:
I would just make the logger field public.

The static method just looks like additional noise to me.